drivers/MbedCRC.h

```diff
@@ -514,11 +514,6 @@ class MbedCRC {
         MBED_STATIC_ASSERT(width <= 32, "Max 32-bit CRC supported");
 
 #if DEVICE_CRC
-        if (POLY_32BIT_REV_ANSI == polynomial) {
```

Comment:
The current code stops the HAL version being called for that case by returning early.

Not 100% clear on why this polynomial needs to be special-cased here, rather than dealt with in the HALs' `hal_crc_is_supported`, but it would make more sense just to make the entire `#if DEVICE_CRC` block have an extra `if (polynomial != POLY_32BIT_REV_ANSI)` test around it. This isn't a good use of early return.

Reply:
Yes I got that it returns early but there is no comments on why it is necessary for `POLY_32BIT_REV_ANSI` or even if it is necessary.

If it is not necessary and the polynomial is dealt with by `hal_crc_is_supported` then we do not need the piece of code removed.

Reply:
I think the point is that the API has been hacked so that polynomial `0xEDB88320` is treated specially elsewhere in the software code. It doesn't actually run the same generic algorithm, so passing polynomial `0xEDB88320` to the HW API won't achieve the same effect unless they're also all modified to treat `0xEDB88320` as a magic number, which presumably they haven't been.

Reply:
Looking at it, I think the issue is that default constructor uses magic values of the `polynomial` template parameter to infer other parameters like reflection.

I think the previous PR tried to use a new magic polynomial value to generate the reversal, but broke what the polynomial parameter meant.

I think this reverse CRC would have been accurately described by

    MbedCRC<POLY_32BIT_ANSI, 32> normal_crc(); // equivalent to (~0, ~0, true, true)
    MbedCRC<POLY_32BIT_ANSI, 32> reverse_crc(~0, ~0, false, false);

without the previous PR.

If we want `MbedCRC<POLY_32BIT_ANSI_REV, 32>` to be able to invoke hardware, we'd need a few bits that did `polynomial == POLY_32_BIT_ANSI_REV ? POLY_32_BIT : polynomial` every time we went to pass it to the hardware.

Reply:
I've just triple-checked what it's doing. `POLY_32BIT_REV_ANSI` actually computes **exactly the same CRC** as `POLY_32BIT_ANSI`... It's not even reversing the output order or anything. (The constructor shows reversal, but the special-case code ignores those parameters).

All it seems to be is expressing the polynomial itself in opposite-order form. Which doesn't change the result.

Reply:
@deepikabhavnani and @rajkan01 could you please comment on the above?

Reply:
I'm going to be looking at the whole thing further; there are general issues with the way the HW/SW selection is done, and image size. And I'll look at what to do with POLY_32BIT_REV_ANSI.

So this PR probably wants to just focus on eliminating the warning for now.

I think the safest change, not affecting functionality, is to wrap the entire hw check block up, leaving the switch statement to handle the REV case.

    #if DEVICE_CRC
    if (POLY_32BIT_REV_ANSI != polynomial) {
        crc_mbed_config_t conf;
        ...
        if (hal_crc_is_supported(&config)) {
            _mode = HARDWARE;
            return;
        }
    }
    #endif

Reply:
https://github.com/ARMmbed/mbed-os/pull/11478#issuecomment-533046223